File: plot_nonadaptive.py
import matplotlib.pyplot as plt
import os
import numpy as np
import itertools
from matplotlib.ticker import FuncFormatter, MaxNLocator


def main():
    epochs = 3000
    nights = 10
    capacity = 10
    numTrials = 20
    maxAgents = 100
    # date = ("2017-11-03_08-14-15") # discount = 0.9
    date = ("2017-11-03_10-02-19") # discount = 0

    numDisabled = [0, 20, 50, 60, 70, 90]
    baseResultsPath = os.path.join("build", "Results", date, "MultiNightBarQ", "non-adaptive")

    paths = map(lambda x: os.path.join(baseResultsPath, str(maxAgents)+"_agents", str(x)+"_disabled"), numDisabled)

    dataDict = {}

    # loop through each variation
    for i in range(len(paths)):
        nDisabled = numDisabled[i]
        path = paths[i]

        csvFname = "results.csv"
        
        trialFolders = ["trial_"+str(i) for i in range(numTrials)]
        # get first CSV to format data array
        trial0 = np.genfromtxt(os.path.join(path, "trial_0", csvFname), delimiter=",")
        data = np.zeros((len(trial0), numTrials))

        for i in range(numTrials):
            trialData = np.genfromtxt(os.path.join(path, trialFolders[i], csvFname), delimiter=",")
            data[:,i] = trialData[:,1]

        # compute Mean and std deviation and store in numpy array
        meanStd = np.zeros((len(trial0), 3))
        meanStd[:,0] = trial0[:,0]
        meanStd[:,1] = np.mean(data, axis=1)
        meanStd[:,2] = np.std(data, axis=1)

        # print data

        # associate array with number of agents not learning in dictionary
        dataDict[nDisabled] = meanStd

    # Plot Performance vs Num Epochs for each variation
    numplots = len(numDisabled)
    # colormap = plt.cm.gist_ncar
    plt.style.use('ggplot')
    # plt.gca().set_prop_cycle(plt.cycler('color', plt.cm.jet(np.linspace(0, 1, numplots))))

    # colors = ["r", "g", "b", "colors", "m", "k"]
    markers = itertools.cycle(('o', 'v', 'x', 's', 'p', '^', '<', '>'))
    linestyles = itertools.cycle(('-', '--', '-.', ':'))

    ax = plt.gca()
    i = 0
    increment = 200
    maxEpoch = 3000
    for key in numDisabled:
        value = dataDict[key]
        x_axis = value[:,0][:maxEpoch:increment]
        y_axis = value[:,1][:maxEpoch:increment]
        errors = value[:,2][:maxEpoch:increment]
        # color = next(ax._get_lines.color_cycle)
        # plt.errorbar(x_axis, y_axis, errors, linestyle='solid', marker=markers.next(), markerfacecolor=color, markeredgecolor=color, c=color, label=str(key), mew=5.0)
        plt.errorbar(x_axis, y_axis, errors, linestyle=linestyles.next(), marker=markers.next(), label=str(key), mew=5.0)
        i+=1

    handles, labels = ax.get_legend_handles_labels()
    box = ax.get_position()
    ax.set_position([box.x0, box.y0, box.width * 0.8, box.height])
    ax.get_yaxis().set_tick_params(which='both', direction='out')
    # major ticks every 20, minor ticks every 5                                      
    major_ticks = np.arange(0, 101, 20)                                              
    minor_ticks = np.arange(0, 101, 5)

    # ax.set_yticklabels(major_ticks,fontsize=20)                                               
                                       
    # ax.set_yticks(major_ticks)                                                       
    # ax.set_yticks(minor_ticks, minor=True) 
    # ax.grid(which='both')                                                            

    # or if you want differnet settings for the grids:                               
    # ax.grid(which='minor', alpha=0.2)                                                
    # ax.grid(which='major', alpha=0.5)   

    plt.yticks(range(0,100,10))
    plt.title("Performance vs Number of Epochs for " + str(nights) + " Nights of " + str(capacity) + " Capacity with " + str(maxAgents) + " Agents")
    plt.xlabel("Number of Epochs")
    plt.ylabel("Performance (max 100)")
    plt.ylim([0,110])
    plt.yticks(np.arange(0, 110, 10))
    ax.legend(handles, labels, loc='center left', bbox_to_anchor=(1, 0.5), title="Number of Agents\nNot Learning")
    plt.show()  



if __name__ == '__main__':
    main()

File: plot/plot_nonadaptive.m
function plot_nonadaptive(varargin)
varargin
numAgents = 100;
legendLoc = 'SouthEast';
if length(varargin)==1
    numAgents = varargin;
elseif length(varargin)==2
    numAgents = varargin{1};
    legendLoc = varargin{2};
end

figure;
epochs = 3000;
nights = 10;
capacity = 10;
numTrials = 20;
% numAgents = 200;

date = '2017-11-03_10-02-19';
if numAgents == 100
    numDisabled = {'0', '20', '50', '70', '90'};

elseif numAgents == 150
    numDisabled = {'0', '20', '50', '80', '100' '120'};

elseif numAgents == 200
    numDisabled = {'0', '50', '80', '100', '150', '170'};

else
    'invalid number of agents'
end
    
% numDisabled = {'0'};

% paths = arrayfun(@(x) strcat('../build/Results/', date, '/MultiNightBarQ/non-adaptive', ...
%     "/", num2str(numAgents),'_agents/',x,'_disabled'),numDisabled);

% paths = arrayfun(@(x) strcat('results_11-8/final_discount0/MultiNightBarQ/non-adaptive', ...
% paths = arrayfun(@(x) strcat('results_11-8/final/MultiNightBarQ/non-adaptive', ...
%     "/", num2str(numAgents),'_agents/',x,'_disabled'),numDisabled);

paths = arrayfun(@(x) strcat('../build/Results/final_discount0/MultiNightBarQ/non-adaptive', ...
    "/", num2str(numAgents),'_agents/',x,'_disabled'),numDisabled);


dataDict = containers.Map();

for i = 1:size(paths,2)
    nDisabled = numDisabled{i};

    path = paths(i)
    
    csvFname = '/results.csv';
    
    trialFolders = arrayfun(@(x) strcat('/trial_',num2str(x)), 0:numTrials-1, 'UniformOutput', false);
    file = strcat(path, '/trial_0', csvFname);
    trial0 = csvread(file);
    data = zeros(size(trial0, 1), numTrials);
    
    for j = 1:numTrials
       trialData =  csvread(strcat(path, trialFolders(j), csvFname));
       data(:,j) = trialData(:,2);
    end
    
    meanStd = zeros(size(trial0, 1), 3);
    meanStd(:,1) = trial0(:,1);
    meanStd(:,2) = mean(data, 2);
    meanStd(:,3) = std(data,0, 2)./sqrt(numTrials);
    
    dataDict(nDisabled) = meanStd;
    
end



markers = ['o'; 'v'; 's'; '^'; 'd'; 'p';'x'];
linestyles = {'-.'; '-'; '--'};
colors = get(gca, 'colororder');
% '-' = baseline
% 'o' for original
set(gcf, 'Position', [1000, 800, 560, 420])
set(gca, 'FontName', 'Times New Roman');
lw = 1;
fs = 14;

increment = 200;
maxEpoch = 2000;
dict_keys = numDisabled;

plotHandles = zeros(length(dict_keys),1);
errHandles = zeros(length(dict_keys),1);
sampleHandles = zeros(length(dict_keys),1);

for i = 1:length(dict_keys)
    key = dict_keys{i};
    value = dataDict(key);
    epochs = value(1:maxEpoch,1);
    means = value(1:maxEpoch,2);
    stderr = value(1:maxEpoch,3);
    
    x_axis = value(1:increment:maxEpoch,1);
    y_axis = value(1:increment:maxEpoch,2);
    errors = value(1:increment:maxEpoch,3);
%     e = plot(x_axis, y_axis,... errors, ...
%         'Marker', markers(mod(i,length(markers))), ...
%         'Linestyle', linestyles{1+mod(i, length(linestyles))} ...
%         );
    
    % Plot line
    
    ls = linestyles{1 + mod(i, length(linestyles))};
    c = colors(i,:);
    mkr = markers(mod(i,length(markers)));
    plotHandles(i) = plot(epochs, means, 'LineStyle', ls, 'LineWidth', lw, 'Color', c);
    hold on
    errHandles(i) = errorbar(x_axis, y_axis, errors, ...
        'LineStyle', 'None', 'Marker', mkr , 'MarkerFaceColor', c, 'Color', c);
    sampleHandles(i) = errorbar(x_axis(1), y_axis(1), errors(1), ...
        'LineStyle', ls, 'Marker', mkr, 'MarkerFaceColor', c, 'Color', c,'LineWidth', lw);
    

end

% title(strcat('Performance vs Epochs for ', num2str(nights), ...
%     ' Nights of ', num2str(capacity), ' Capacity with ', num2str(numAgents), ' Adaptive ', 'Agents'));

xlabel('Epoch', 'FontSize', 14, 'Interpreter', 'latex');
legend(sampleHandles,numDisabled, 'Location', legendLoc, 'Interpreter', 'latex');
ylim([10,100]);

set(gca,'fontname','Times New Roman','FontSize',fs)
grid on 

if numAgents == 100
    ylabel('Performance (max 100)', 'FontSize', fs, 'Interpreter', 'latex');
    savefig('bar_nonadaptive_100agents.fig')
    export_fig(gcf, 'bar_nonadaptive_100agents.pdf', '-trans');

elseif numAgents == 150
    ylabel('Performance (max 90)', 'FontSize', fs, 'Interpreter', 'latex');
    savefig('bar_nonadaptive_150agents.fig')
    export_fig(gcf, 'bar_nonadaptive_150agents.pdf', '-trans');

elseif numAgents == 200
    ylabel('Performance (max 90)', 'FontSize', fs, 'Interpreter', 'latex');
    savefig('bar_nonadaptive_200agents.fig')
    export_fig(gcf, 'bar_nonadaptive_200agents.pdf', '-trans');

else
'invalid number of agents, cant export_fig'

end
end